// react-router5/dist/es/routeNode.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import { Component, createElement } from 'react';
import { getDisplayName } from './utils';
import PropTypes from 'prop-types';
import { shouldUpdateNode } from 'router5-transition-path';

function routeNode(nodeName) {
    return function routeNodeWrapper(RouteSegment) {
        var RouteNode = function (_Component) {
            _inherits(RouteNode, _Component);

            function RouteNode(props, context) {
                _classCallCheck(this, RouteNode);

                var _this = _possibleConstructorReturn(this, (RouteNode.__proto__ || Object.getPrototypeOf(RouteNode)).call(this, props, context));

                _this.router = context.router;
                _this.routeState = {
                    previousRoute: null,
                    route: _this.router.getState()
                };
                _this.mounted = false;

                if (typeof window !== 'undefined') {
                    var listener = function listener(_ref) {
                        var route = _ref.route,
                            previousRoute = _ref.previousRoute;

                        if (shouldUpdateNode(nodeName)(route, previousRoute)) {
                            _this.routeState = {
                                previousRoute: previousRoute,
                                route: route
                            };
                            if (_this.mounted) {
                                _this.forceUpdate();
                            }
                        }
                    };
                    _this.unsubscribe = _this.router.subscribe(listener);
                }
                return _this;
            }

            _createClass(RouteNode, [{
                key: 'componentDidMount',
                value: function componentDidMount() {
                    this.mounted = true;
                }
            }, {
                key: 'componentWillUnmount',
                value: function componentWillUnmount() {
                    if (this.unsubscribe) {
                        this.unsubscribe();
                    }
                }
            }, {
                key: 'render',
                value: function render() {
                    var props = this.props,
                        router = this.router;
                    var _routeState = this.routeState,
                        previousRoute = _routeState.previousRoute,
                        route = _routeState.route;

                    var component = createElement(RouteSegment, _extends({}, props, {
                        router: router,
                        previousRoute: previousRoute,
                        route: route
                    }));

                    return component;
                }
            }]);

            return RouteNode;
        }(Component);

        RouteNode.contextTypes = {
            router: PropTypes.object.isRequired
        };

        RouteNode.displayName = 'RouteNode[' + getDisplayName(RouteSegment) + ']';

        return RouteNode;
    };
}

export default routeNode;
