// react-router5/dist/commonjs/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.RouteNode = exports.Route = exports.RouteProvider = exports.Link = exports.withRouter = exports.withRoute = exports.RouterProvider = exports.routeNode = exports.BaseLink = undefined;

var _BaseLink = require('./BaseLink');

var _BaseLink2 = _interopRequireDefault(_BaseLink);

var _routeNode = require('./routeNode');

var _routeNode2 = _interopRequireDefault(_routeNode);

var _RouterProvider = require('./RouterProvider');

var _RouterProvider2 = _interopRequireDefault(_RouterProvider);

var _withRoute = require('./withRoute');

var _withRoute2 = _interopRequireDefault(_withRoute);

var _withRouter = require('./withRouter');

var _withRouter2 = _interopRequireDefault(_withRouter);

var _RouteProvider = require('./RouteProvider');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Link = (0, _withRoute2.default)(_BaseLink2.default);

exports.BaseLink = _BaseLink2.default;
exports.routeNode = _routeNode2.default;
exports.RouterProvider = _RouterProvider2.default;
exports.withRoute = _withRoute2.default;
exports.withRouter = _withRouter2.default;
exports.Link = Link;
exports.RouteProvider = _RouteProvider.RouteProvider;
exports.Route = _RouteProvider.Route;
exports.RouteNode = _RouteProvider.RouteNode;
